### PR TITLE
raftstore: Inherit raftdb-path when upgrading to raft-engine

### PR DESCRIPTION
**Pull request by Smityz**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/15771

What's Changed:

```commit-message
Inherit raftdb-path when upgrading to raft-engine, 
and support to custom set raft-engine dir when upgrading
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

**Comment by Smityz** [comment]
I added some tests so this PR looks so big. Actually, there is only a little change in the logic of detecting if raft data exists.
@tabokie @tonyxuqqi @disksing

File: components/tikv_util/src/config.rs [hunk 1]
```diff
@@ -1569,18 +1570,66 @@ impl RaftDataStateMachine {
...
+                        let path = entry.path();
+                        if path.is_file() && path.extension() == Some("raftlog".as_ref()) {
+                            return true;
+                        }
+                    }
+                    Err(e) => {
+                        warn!("Scan raft engine dir failed"; "path" => %raftengine_path.display(), "err" => ?e)
+                    }
+                };
+            }
```
Comment (LykxSassinator) [inline components/tikv_util/src/config.rs:1619]:
```suggestion
    pub fn raftengine_exists(path: &Path) -> bool {
        if !dir_exists(path) {
            return false;
        }
        fs::read_dir(path).unwrap().any(|entry| {
            if let Ok(e) = entry {
                let p = e.path();
                p.is_file() && p.extension().map_or(false, |ext| ext == "raftlog")
            } else {
                false // No need to warn
            }
        })
```

File: components/tikv_util/src/config.rs [hunk 2]
```diff
@@ -1569,18 +1570,66 @@ impl RaftDataStateMachine {
...
-        fs::read_dir(path).unwrap().next().is_some()
+        false
+    }
+
+    pub fn raftdb_exist(raftdb_path: &Path) -> bool {
+        if raftdb_path.is_dir() && raftdb_path.join("CURRENT").exists() {
+            return true;
+        }
+        false
+    }
```
Comment (LykxSassinator) [inline components/tikv_util/src/config.rs:1629]:
```suggestion
    pub fn raftdb_exists(path: &Path) -> bool {
        if !dir_exists(path) {
            return false;
        }
        let current_file_path = path.join("CURRENT");
        current_file_path.exists() && current_file_path.is_file()
    }
```

File: components/tikv_util/src/config.rs [hunk 3]
```diff
@@ -1569,18 +1570,66 @@ impl RaftDataStateMachine {
...
+
+    pub fn raft_data_exists(path: &Path) -> bool {
+        Self::raftengine_exist(path) || Self::raftdb_exist(path)
```
Comment (LykxSassinator) [inline components/tikv_util/src/config.rs:1632]: IMO, it's redundant to rename `data_exists` to `raft_data_exists`, as the naming of this class already contains this.
```suggestion
    pub fn data_exists(path: &Path) -> bool {
        Self::raftengine_exists(path) || Self::raftdb_exists(path)
```

File: components/tikv_util/src/config.rs [hunk 4]
```diff
@@ -1569,18 +1570,66 @@ impl RaftDataStateMachine {
...
+            info!("Path not exists"; "path" => %remove_path.display());
+            return;
+        }
+        if !remove_path.is_dir() {
+            info!("Path is not a directory, so remove directly"; "path" => %remove_path.display());
+            Self::must_remove(remove_path);
+            return;
+        }
+        if !retain_path.starts_with(remove_path) {
+            info!("Retain path is not under the remove path, so remove directly"; "Retain path" => %retain_path.display(), "remove path" => %remove_path.display());
```
Comment (LykxSassinator) [inline components/tikv_util/src/config.rs:1585]:
```suggestion
            info!("Removing directory as retain path is not under remove path"; "retain path" => %retain_path.display(), "remove path" => %remove_path.display());
```

**Comment by LykxSassinator** [comment]
@Smityz please check and fix the clippy errors.

**Review by LykxSassinator — APPROVED** [review]
Rest LGTM.
/cc @tabokie

**Pull request merged**